doc/reference/trd-syscalls.md

```diff
@@ -539,8 +587,60 @@ are 12:
 | 10              | Set the start of the process stack                      | Success          |
 | 11              | Set the start of the process heap                       | Success          |
 
-The success return variant is Memop class system call specific. All
-Memop class system calls have a _Failure_ failure type.
+The success return variant is Memop class system call specific and
+specified in the table above. All Memop class system calls have a
+_Failure_ failure type.
+
+4.7 Exit (Class ID: 6)
+--------------------------------
+
+The Exit system call class is how a userspace process terminates. 
+Successful calls to Exit system calls do not return.
+
+There are two Exit system calls:
+  - `exit-terminate`
+  - `exit-restart`
+
+The first call, `exit-terminate`, terminates the process and tells the
+kernel that it may reclaim and reallocate the process as well as all of its
+resources. Usually this indicates that the process has completed
+its work. 
+
+The second call, `exit-restart`, terminates the process and tells the kernel
+that the application would like to restart if possible. If the kernel 
+restarts the application, it MUST assign it a new process identifier. The
+kernel MAY reuse existing process resources (e.g., RAM regions) or MAY
+allocate new ones.
+
+The register arguments for Exit system calls are as follows. The registers
+r0-r3 correspond to r0-r3 on CortexM and a0-a3 on RISC-V.
+
+| Argument          | Register |
+|-------------------|----------|
+| Exit identifer    | r0       |
+| Completion code   | r1       |
+
+The exit identifier specifies which call is invoked.
+
+| System call     | Exit identifier value |
+|-----------------|-----------------------|
+| exit-terminate  |                     0 |
+| exit-restart    |                     1 |
+
+The difference between `exit-terminate` and `exit-restart` is what behavior
+the application asks from the kernel. With `exit-terminate`, the application
+tells the kernel that it considers itself completed and does not need to run
+again. With `exit-restart`, it tells the kernel that it would like to be
+rebooted and run again. For example, `exit-terminate` might be used by a 
+process that stores some one-time data on flash, while `exit-restart` might
+be used if the process runs out of memory.
+
+The completion code is an unsigned 32-bit number which indicates status. This
+information can be stored in the kernel and used in management or policy decisions.
+The definition of these status codes is outside the scope of this document.
```

Comment:
Why not at least say a completion code of 0 is success?

Reply:
Because that's an aspect of POSIX semantics which I'm not certain we want to borrow. We very well might, but we haven't fully thought this through. Rather than block this document on a design for process exit codes, I thought it better to just defer to another document. Nothing depends on or uses this yet, so binding it so early seems like a mistake.

Reply:
We should bind it at some point, otherwise different apps are going to return different values and it will become fragmented.

I don't think returning 0 for success is just a POSIX thing, doesn't Windows and everything else do that as well?

Just to clarify I'm not saying specify the meanings, just something along the lines of negative numbers mean failure, 0 means success (or something like that).

Reply:
Also, we could copy the low level debug meanings here and then remove it and replace it with this instead.

Reply:
I kinda think the low level debug alert codes should have been runtime-specific from the start, so I'm not sure I agree.

Reply:
I agree we should bind it at some point (even very soon). I just think that this isn't the place to do it. It's going to be part of discussing, designing, and specifying the lifetime of a process in Tock.